README.md
# Harden FreeBSD 

![](images/harden-freebsd-logo.jpg)

FreeBSD officially defaults to [Permanently Insecure Mode](https://man.freebsd.org/cgi/man.cgi?securelevel). This script will duplicate all the hardening settings run by `/usr/libexec/bsdinstall/hardening` and much more. Any directive can be set and re-set with a customizable `settings.ini` for administering, tuning, and easier jail management. All existing entries in all confs will remain untouched unless they are modified in the settings file.

This script is also targeted to new users of FreeBSD so that they may leverage years of security contributions by the entire BSD community across all spectra, implemented on thier system in seconds.

Each of the security settings was researched, assessed, and chosen as a set of mitigations for maximizing threat reduction while minimizing restriction of system capability and availability.


## New Features in 2.0.1
* Now sets `/boot/loader.conf` in order to load kernel security features early in initialization
* New security settings
* Additional software included to verify the implementation

*Full [Changelog](Changelog.md)*


## Addtional Software
* Scripts included to verify the implementation 
    * Kernel vulnerablity diagnosis provided by [Stéphane Lesimple's](https://github.com/speed47) spectre-meltdown-checker
        * `chmod 750 spectre-meltdown-checker.sh`
        * You should only be left with the MCEPSC, Machine Check Exception on Page Size Change Vulnerability, [CVE-2018-12207](https://www.freebsd.org/security/advisories/FreeBSD-SA-19:25.mcepsc.asc)
    * MMAP, MProtect vulnerability diagnosis provided by [u/zabolekar](https://www.reddit.com/r/BSD/comments/10isrl3/notes_about_mmap_mprotect_and_wx_on_different_bsd/)
        * `cc mmap_protect.c` 
        * `./a.out`
        * You should have two successes


## Features

* Makes backups of `rc.conf`, `sysctl.conf`, `login.conf`, and `loader.conf` on first run
* Sets passwords to blowfish encryption
* Sets passwords to expire at 120 days
* Disables sendmail completely
* Removes `other` write permissions from key system files and folders
* Allows only root for `cron` and `at`
* Primitive flag verification catches simple errors
* Modularizable within other tools
* Automate any shell script
* System Logging to `/var/log/messages` and Script Logging to `/var/log/harden-freebsd.log`
* Pretty prints color output of script execution to console while running


## Requirements
* FreeBSD 13.1, 13.2
* Python 3.9.16


## Installation

**WARNING: Once kernel level 1 is set by this script, you will not be able to modify these confs again with this script until it is set to -1 and rebooted!**

* Set `kernlevel = -1` if you want to test various setting groups with your applications and network
* Customize `settings.ini`  to whatever is needed, the script will change the directive to your flag
* Set permissions `chmod 750 harden-freebsd.py` to prevent shell injection from another account or process
* Set permissions `chmod 640 settings.ini` to prevent shell injection from another account or process
* No `settings.ini` section can be entirely commented out nor be completely empty
* `./harden-freebsd.py`


## Customization

#### 64bit vs 32bit
Most tunable mitigations for 64bit are already included by default in FreeBSD 13.1 so 32bit directives were included for coverage. I can see no affect from setting the 32bit mitigations on 64bit systems, they are simply ignored. For clarity on unknown hardware, hardware mode, VM, or cloud use the following commands:
* CPU: `sysctl hw.model hw.machine hw.ncpu`
* Bits: `getconf LONG_BIT`


#### Backups

The very first time the script is run it will make copies of `rc.conf`, `sysctl.conf`, `login.conf`, and `loader.conf` named `rc.conf.original` etc. If you've already done this yourself you may want to rename or move those files.

If you would like you can set `settings.ini` section `[SCRIPT]`option `first_run` to `True` with capital `T` to make new backups at any time after you've renamed the original backups or the script will overwrite them.

#### Verification

To err on the safe side, the script does primitive verification of the confs flags. Check the log for any validation failures and rewrite the regular expression or make a new check.

* For `/etc/sysctl.conf` the script checks for no quotes
* For `/boot/loader.conf` the script strictly verifies syntax from man and `/boot/defaults/loader.conf` syntax
    * All directives in these sister confs must be in quotes

If you do get stuck in read-only single-user mode and need to correct a configuration file then use:

```sh
zfs set readonly=false zroot
zfs mount -a
```
#### Chmod-ability

The set of files needed to be secure changed and changed throughout testing and so it ended up as a shell command but an error checked function was provided for the administrator programmer to use instead of appending to the long list in `settings.ini` section `[FILESEC]` if you wish or to work with other software.

Those files are:

`etc/ftpusers /etc/group /etc/hosts /etc/hosts.allow /etc/hosts.equiv /etc/hosts.lpd /etc/inetd.conf /etc/login.access /etc/login.conf /etc/newsyslog.conf /etc/rc.conf /etc/ssh/sshd_config /etc/sysctl.conf etc/syslog.conf /etc/ttys /etc/crontab /usr/bin/crontab /usr/bin/at /usr/bin/atq /usr/bin/atrm /usr/bin/batch /var/log`

#### Secure Password Settings

The newly applied settings will not take affect until you reset your password.

## Automatic Jail Lockdown/Management Strategies

1. Set the correct paths to jailed confs in `harden-freebsd.py` lines 32-38 and run for each jail.
2. Copy software to `/root` and have jail start this script at reboot and all settings will be updated upon next reboot. To update all jails simply copy `settings.ini` with your own copy script to all appropriate locations for uptake. 

`crontab -e`

`@reboot /path/to/harden-freebsd.py`

3. Have all jails pointing to the same rc script via `exec.start` and set paths in the script pointing to the same location modified by the script paths. 
4. Add new jail specific entires to `settings.ini [SYSTEM]` section for sysctl.conf udpate
   - `security.jail.* = 0`
5. Use mutiple copies of the script and settings.ini for each jail
6. Put it in your template


## Setting Descriptors
**Startup**
* `kern_securelevel_enable = "YES"`
    * Enable access to other than permanently insecure modes
* Disable Sendmail
* `syslogd_flags="-ss"`
    * Disallow syslogd to bind to a network socket
* `clear_tmp_enable = "YES"`
    * Clear the /tmp directory on reboot
* `icmp_drop_redirect="YES"`
    * Disallow redirection of ICMP (ping, echo)
* `inetd_enable = "NO"`
    * Disallow Network File System to share directories over the network
* `portmap_enable = "NO"`
    * Disallow portmapping since Network File Systems is disallowed
* `update_motd = "NO"`
    * Disallow computer system details from being added to /etc/motd on system reboot

**System**

* `kern.securelevel = 1` [(*)](https://man.freebsd.org/cgi/man.cgi?securelevel)
    * The system immutable and system append-only flags may
	   not be turned off; disks for	mounted	file systems, /dev/mem and
	   /dev/kmem may not be	opened for writing; /dev/io (if	your platform
	   has it) may not be opened at	all; kernel modules (see kld(4)) may
	   not be loaded or unloaded.  The kernel debugger may not be entered
	   using the debug.kdb.enter sysctl.  A	panic or trap cannot be	forced
	   using the debug.kdb.panic, debug.kdb.panic_str and other sysctl's.
* `security.bsd.see_other_uids = 0`
    * Disallow users from seeing information about processes that are being run by another user (UID)
* `security.bsd.see_other_gids = 0` [(*)](https://docs.freebsd.org/en/books/handbook/mac/#mac-policies)
    * Disallow users from seeing information about processes that are being run by another group (GID)
* `security.bsd.see_jail_proc = 0` (Sysctl MIB Entry `sysctl -a | grep security.bsd`)
    * Disallow non-root users from seeing processes in jail
* `security.bsd.unprivileged_read_msgbuf = 0` (Sysctl MIB Entry `sysctl -a | grep security.bsd`)
    * Disallow non-root users from reading system message buffer
* `kern.randompid = 107` [(*)](https://wiki.freebsd.org/DevSummit/201308/Security)
    * Force kernel to randomize process ID's using above salt value instead of sequential
* `net.inet.ip.random_id = 1`
    * Randomize IP packet ID
* `net.inet.ip.redirect = 0` 
    * Disallow ICMP host redirects
* `net.inet.tcp.always_keepalive = 0`
    * Disallow keeping open idle TCP connections
* `net.inet.tcp.blackhole = 2` +(UDP)[(*)](https://man.freebsd.org/cgi/man.cgi?query=blackhole)
    * Packets that are received on a closed port will not initiate a reply
* `net.inet.tcp.path_mtu_discovery = 0` [(*)](https://man.freebsd.org/cgi/man.cgi?query=tcp&sektion=4)
    * Disallows TCP to determine the minimum MTU size on any network that is currently in the path between two hosts
* `net.inet.icmp.drop_redirect = 1`
    * Pairs with rc.conf startup, as once enabled, it is then set
* `hw.mds_disable = 1` [(*)](https://www.kernel.org/doc./html/latest/arch/x86/mds.html)
    * Enable Microarchitectural Data Sampling Mitigation version `VERW`
    * Change value to `3` (AUTO) if using a Hypervisor without MDS Patch
* `hw.spec_store_bypass_disable = 1` [(*)](https://handwiki.org/wiki/Speculative_Store_Bypass)
    * Disallow Speculative Bypass used by Spectre and Meltdown
* `kern.elf64.allow_wx = 0` [(*)](https://www.ibm.com/docs/en/aix/7.2?topic=memory-understanding-mapping)
    * Disallow write and execute for shared memory


**Kernel**
* `security.bsd.allow_destructive_dtrace = "0"`
    * Disallow DTrace to terminate proccesses
    * Test DTrace hardening: Using all 3 commands should result in `Permission denied` or `Destructive actions not allowed`:
    * `dtrace -wn 'tcp:::connect-established { @[args[3]->tcps_raddr] = count(); }'`
    * `dtrace -wqn tick-1sec'{system("date")}'`
    * `dtrace -qn tick-1sec'{system("date")}'`
* `hw.ibrs_disable = "3"` [(*)](https://wiki.freebsd.org/SpeculativeExecutionVulnerabilities)
    * Prevent Spectre and Meltdown CPU Vulnerabilities, 3 for AUTO
* `kern.elf32.aslr.stack = "3"` [(*)](https://wiki.freebsd.org/AddressSpaceLayoutRandomization)
    * Address space layout randomization is used to increase the difficulty of performing a buffer overflow attack
    * 64bit is enabled by default in 13.2 so you can set this to 0 for 64bit processors or remove
* `kern.elf32.aslr.pie_enable = "1"`
    * Enable ASLR for Position-Independent Executables (PIE) binaries
* `cpu_microcode_load = "NO"`
    * Disallow automatic CPU Microcode updates


## License Summary

Non-Commercial usage, retain and forward author and license data. Modify existing code as needed up to 25% while allowing unlimited new additions. The Software may use or be used by other software.


## Security Guidelines

Since this Software uses shell commands it is required to place it in a secure directory with permissions on the **parent** directory to have no permissions for `other` /all/world group to write or *execute* and **no network access**. 

Please follow [these guidelines](/docs/SECURITY.md) should you find a vulnerability not addressed in the audit.


## Statement of Security: 

* **Risk** - Low
* **Impact** - Medium

This script has no networking, accesses no sockets, and uses only standard libraries.

Although this script is using `subprocess.run(shell=True)` the only possibility of shell injection is from the paths customized by the Licensee or unauthorized access to the filesystem the script resides on in order to perform unauthorized modifications to `settings.ini`or the Software which is not a vulnerability of the Software. 


### Latest Development Version

[Quadhelion Engineering Code Repository](https://got.quadhelion.engineering)



![quadhelion engineering logo](images/quadhelionEngineeringIcon.jpg)

Changelog
June 23, 2023
64bit ASLR setting was reverted back to 32bit enable as 64bit is enabled by default.

mmap_protect.c
#include <sys/mman.h>
#include <stdio.h>

// Author: u/zabolekar
// URL: https://www.reddit.com/r/BSD/comments/10isrl3/notes_about_mmap_mprotect_and_wx_on_different_bsd/
// Customized by Quadhelion Engineering
// Compile it with cc mmap_mprotect.c, run it with ./a.out

int main()
{
    void* p = mmap(NULL, 1, PROT_WRITE|PROT_EXEC, MAP_ANON|MAP_PRIVATE, -1, 0);

    if (p == MAP_FAILED)
    {
        perror("\n*********************\033[38;5;76m Success \033[0;0m*************************\n Permission denied for shared memory map write and execute\n*******************************************************\n");
    }
    else
    {
        puts("VULNERABLE: Writable and executable memory mapped successfully");
        munmap(p, 1);
    }

    p = mmap(NULL, 1, PROT_WRITE, MAP_ANON|MAP_PRIVATE, -1, 0);
    if (p == MAP_FAILED)
    {
        perror("Map writable memory");
    }
    else
    {
        puts("\n * Writable memory mapped successfully");

        if (mprotect(p, 1, PROT_EXEC))
            perror("Can't make writable memory executable");
        else
            puts(" * Preparing writable memory for execution");

        if (mprotect(p, 1, PROT_WRITE|PROT_EXEC))
            perror("\n*********************\033[38;5;76m Success \033[0;0m*************************\n Can't make shared memory writable and executable\n*******************************************************\n");
        else
            puts("VULNERABLE: Shared memory successfully made writable and executable");

        munmap(p, 1);
    }
    puts("\n");
}

harden-freebsd.py
#!/usr/bin/env python

""" 
Harden FreeBSD system perms, settings.
Set and reset rc, sysctl, login, confs; set file perms, run shell commands
No arguments, only uses settings.ini file in the same directory.
Example: sendmail_enable = "NONE"
"""

__author__ = "Elias Christopher Griffin"
__url__ = "https://www.quadhelion.engineering"
__license__ = "QHELP-OME-NC-ND-NAI"
__copyright__ = "https://www.quadhelion.engineering/QHELP-OME-NC-ND-NAI.html"
__version__ = "2.0.2"
__date__ = "06/27/2023"
__email__ = "elias@quadhelion.engineering"
__status__ = "Production"



from pathlib import Path
from datetime import datetime
import os, re, subprocess, syslog, configparser, shutil, sys

_date = datetime.now()
date_time = _date.strftime("%m/%d/%Y, %H:%M")

config = configparser.ConfigParser()
config.read('settings.ini')


harden_freebsd_log = Path("/var/log/harden-freebsd.log")
rc_conf = Path("/etc/rc.conf")
sysctl_conf = Path("/etc/sysctl.conf")
loader_conf = Path("/boot/loader.conf")
login_conf = Path("/etc/login.conf")
cron_access = Path("/var/cron/allow")
at_access = Path("/var/at/at.allow")


backup_suffix = ".original"
rc_backup = rc_conf.with_name(rc_conf.name + backup_suffix)
sysctl_backup = sysctl_conf.with_name(sysctl_conf.name + backup_suffix)
loader_backup = loader_conf.with_name(loader_conf.name + backup_suffix)
login_backup = login_conf.with_name(login_conf.name + backup_suffix)



def exception_handler(func):
    def intake(*args, **kwargs):
        try:
            func(*args, **kwargs)
        except PermissionError as e:
            print(f"\n******************\033[38;5;1m Permissions \033[0;0m************************\n")
            print(f"Insufficient permissions {e}")
            print(f"*******************************************************\n")
        except OSError as e:
            print(f"\n********************\033[38;5;1m Locked \033[0;0m**************************\n")
            print(f"Perhaps file is busy, locked, process blocked, or raced:\n") 
            print(f"{e}")
            print(f"*******************************************************\n")
    return intake



# Write to either "syslog" (/var/log/messages) or "script" (/var/log/harden-freebsd.log)
@exception_handler
def writeLog(log_type, content):
    harden_freebsd_logwriter = open(harden_freebsd_log, "a")
    syslog.openlog("LOG_INFO")
    if log_type == "script":
        harden_freebsd_logwriter.writelines(content + os.linesep)
    elif log_type == "syslog":
            syslog.syslog(1, content)
    else:
        print(f"*******************************************************\n")
        print(f"\033[38;5;63m LOG: \033[0;0m {content}")
        print(f"*******************************************************\n")



# Make *.original backups of all files only once
if config['SCRIPT']['first_run'] == "True":
    try:
        harden_freebsd_log.touch()
        cron_access.touch()
        at_access.touch()
        rc_backup.write_bytes(rc_conf.read_bytes())
        sysctl_backup.write_bytes(sysctl_conf.read_bytes())
        loader_backup.write_bytes(loader_conf.read_bytes())
        login_backup.write_bytes(login_conf.read_bytes())
    except FileNotFoundError as e:
        error_path = Path(e.filename)
        print(f"\n********************\033[38;5;1m File Not Found \033[0;0m*******************")
        print(f"Filename: {error_path.parts}")
        print("*******************************************************\n")
    except PermissionError as e:
        print(f"\n******************\033[38;5;1m Permissions \033[0;0m************************\n")
        print(f"Insufficient permissions {e}")
        print(f"*******************************************************\n")
    except OSError as e:
        print(f"\n********************\033[38;5;1m Locked \033[0;0m**************************\n")
        print(f"Perhaps file is busy, locked, process blocked, or raced:\n") 
        print(f"{e}")
        print(f"*******************************************************\n")
    else:
        writeLog("syslog", "System file backups complete")
        with open('settings.ini', 'w') as configfile:
            config.set('SCRIPT', 'first_run', 'False')
            config.write(configfile)
        print(f"\n*********************\033[38;5;76m Success \033[0;0m*************************")
        print(f"\033[38;5;75mCreated: \033[0;0m")
        print(f" {cron_access.name}, {at_access.name} \n")
        print(f"\033[38;5;75mBackups Made: \033[0;0m")
        print(f" {rc_backup.name}, {sysctl_backup.name}")
        print(f" {login_backup.name}, {loader_backup.name} \n")
        print(f"*******************************************************\n")



# Read the system file content
try:
    rc_content = rc_conf.read_text(encoding="utf-8")
    sysctl_content = sysctl_conf.read_text(encoding="utf-8")
    login_content = login_conf.read_text(encoding="utf-8")
    loader_content = loader_conf.read_text(encoding="utf-8")
except FileNotFoundError as e:
    error_path = Path(e.filename)
    writeLog("script", "Error finding file " + error_path)
    print(f"\n*******************\033[38;5;1m File Not Found \033[0;0m********************")
    print(f"Filename: {error_path.name}")
    print(f"Directories used: {error_path.parts}\n")
    print("*******************************************************\n")
except PermissionError as e:
    print(f"\n******************\033[38;5;1m Permissions \033[0;0m************************\n")
    print(f"Permission to read/append {e}")
    print(f"{os.stat(rc_conf)}{os.linesep}")
    print(f"{os.stat(sysctl_conf)}{os.linesep}")
    print(f"{os.stat(loader_conf)}{os.linesep}")
    print(f"{os.stat(cron_access)}{os.linesep}")
    print(f"{os.stat(at_access)}{os.linesep}")
    print(f"{os.stat(login_conf)}{os.linesep}")
    print(f"*******************************************************\n")
else:
    print(
    f"{os.linesep}*********************\033[38;5;75m Running \033[0;0m*************************{os.linesep}"
    f"Loaded system files for read/append\n"
    f"*******************************************************{os.linesep}"
    )
finally:
    writeLog("syslog", "Hardening in progress")
    print(f"\n********************\033[38;5;75m Info Panel \033[0;0m***********************")
    print(f"Executing {__file__}")
    print(f"Executing {date_time}")
    print(f"*******************************************************\n")
    

# Main working class dealing with rc.conf and sysctl.conf
class Conf:
    def __init__(self, file, setting, flag):
        self.file = file
        self.setting = setting
        self.flag = flag
    
    # Changes the flag from whatever it is currently to flag in settings.ini
    def setConf(self):
        try:
            with open(self.file, 'r+', encoding="us-ascii") as file_content: 
                lines = file_content.readlines()
                for i, line in enumerate(lines):
                    if line.startswith(self.setting):
                        lines[i] = self.setting + "=" + self.flag + os.linesep
                        file_content.seek(0)
                        for line in lines:
                            file_content.write(line)
                            file_content.truncate()
                writeLog("script",  self.setting + " was set to " + self.flag)
        except OSError as e:
            print(f"\n********************\033[38;5;1m Locked \033[0;0m**************************\n")
            print(f"Perhaps file is busy, locked, process blocked, or raced:\n") 
            print(f"{e}")
            print(f"*******************************************************\n")
        else:
            print(f"\033[38;5;208m {self.setting} \033[0;0m changed to\033[38;5;208m {self.flag}\033[0;0m  in\033[38;5;75m {self.file}\033[0;0m ")
    
    # Appends at the end of a file a directive that was not present previously
    def addConf(self):
        try:
            with open(self.file, 'a') as file_content: 
                file_content.write(self.setting + "=" + self.flag + os.linesep)
                writeLog("script", self.setting + "=" + self.flag + " added")
        except OSError as e:
            print(f"\n********************\033[38;5;1m Locked \033[0;0m**************************\n")
            print(f"Perhaps file is busy, locked, process blocked, or raced:\n") 
            print(f"{e}")
            print(f"*******************************************************\n")
        else:
            print(f"\033[38;5;63m {self.setting} \033[0;0m added to \033[38;5;75m{self.file}\033[0;0m ")
    
    # Checks to see if the directive is already in the conf, returns True if present.
    def checkConf(self) -> bool:
        self.found = False
        try:
            with open(self.file, 'r') as file_content:
                lines = file_content.readlines()
                for i, line in enumerate(lines):
                    if line.startswith(self.setting):
                        self.found = True
                    else:
                        pass
            return self.found
        except OSError as e:
            print(f"\n********************\033[38;5;1m Locked \033[0;0m**************************\n")
            print(f"Perhaps file is busy, locked, process blocked, or raced:\n") 
            print(f"{e}")
            print(f"*******************************************************\n") 
    
    # Checks proper flag and equality syntax in rc.conf and sysctl.conf with first-boot 13.2 directives
    # May not work with advanced flags added later
    def verifyConf(self): 
        global conf_directives
        conf_directives = []
        sysctl_conf_verify = re.compile(r'[^\"]') # No quotes
        loader_rc_conf_verify = re.compile(r'^[\"].+[$\"]') # Pair of quotes
        try:
            with open(self.file, 'r+') as file_content:
                lines = file_content.readlines()
                for i, line in enumerate(lines):
                    partitioned_line = line.partition("=")
                    if line.isspace():
                        pass
                    elif line.startswith("#"):
                        pass
                    elif partitioned_line[1] != "=":
                        print(f"\n*******************************************************")
                        print(f"Error at {lines[i]}: No equality operator. Restored original.")
                        print(f"*******************************************************\n")
                        writeLog("script", "No equality operator at line " + lines[i].rstrip() + " in " + self.file.name)
                        self.restoreConf()
                        sys.exit()
                    elif self.file == rc_conf and re.match(loader_rc_conf_verify, partitioned_line[2]) == None:
                        print(f"\n*******************************************************")
                        print(f"Error: {self.flag} not allowed in {lines[i]} in {self.file}. Restored original.")
                        print(f"*******************************************************\n")
                        writeLog("script", "Quote matching error at line " + lines[i].rstrip())
                        self.restoreConf()
                        sys.exit()
                    elif self.file == loader_conf and re.match(loader_rc_conf_verify, partitioned_line[2]) == None:
                        print(f"\n*******************************************************")
                        print(f"Error: {self.flag} not allowed in {lines[i]} in {self.file}. Restored original.")
                        print(f"*******************************************************\n")
                        writeLog("script", "Quote matching error at line " + lines[i].rstrip())
                        self.restoreConf()
                        sys.exit()
                    elif self.file == sysctl_conf and re.match(sysctl_conf_verify, partitioned_line[2]) == None:
                        print(f"\n*******************************************************")
                        print(f"Error: {self.flag} not allowed in {lines[i].rstrip()} in {self.file}. Restored original.")
                        print(f"*******************************************************\n")
                        writeLog("script", "Quote in sysctl.conf at line " + lines[i])
                        self.restoreConf()
                        sys.exit()
                    else:
                        conf_directives.append(line.rstrip())
        except OSError as e:
            print(f"\n********************\033[38;5;1m Locked \033[0;0m**************************\n")
            print(f"Perhaps file is busy, locked, process blocked, or raced:\n") 
            print(f"{e}")
            print(f"*******************************************************\n")
    
    # If syntax verification fails, restore *.originals to prevent boot failure
    # If in single user read-only mode use commands:
    # zfs set readonly=false zroot
    # zfs mount -a
    def restoreConf(self):
        try:
            if self.file == rc_conf:
                shutil.copy(rc_backup, rc_conf)
            elif self.file == sysctl_conf:
                shutil.copy(sysctl_backup, sysctl_conf)
            elif self.file == loader_conf:
                shutil.copy(loader_backup, loader_conf)
        except FileNotFoundError as e:
            error_path = Path(e.filename)
            print(f"\n*******************\033[38;5;1m File Not Found \033[0;0m********************")
            print(f"Filename: {error_path.name}")
            print(f"Directories used: {error_path.parts}\n")
            print("*******************************************************\n")
        else:
            print(f"\n*********************\033[38;5;76m Success \033[0;0m*************************")
            print(f"Files restored")
            print("*******************************************************\n")



# Hardcoded sections as only t e contain flags we can dynamically set and re-set.
# Loops through all directives and sets each
class SetOpts:
    def __init__(self, section):
        self.section = section
        if self.section == "STARTUP":
            file = rc_conf
        elif self.section == "SYSTEM":
            file = sysctl_conf
        elif self.section == "KERNEL":
            file = loader_conf
        else:
            pass

        for opt in config[self.section]:
            value = config[self.section][opt]
            conf_runner = Conf(file, opt, value)
            setting_present = conf_runner.checkConf()
            if setting_present:
                conf_runner.setConf()
                conf_runner.verifyConf()
            else:
                conf_runner.addConf()
                conf_runner.verifyConf()
        


# Run shell commands for named section. Will error if sent setting.ini sections that have no shell commands.
def shellCommand(section):
    try:
        for opt in config[section]:
            value = config[section][opt]
            command_result = subprocess.run([value], shell=True, timeout=0.7)
    except subprocess.CalledProcessError as e:
        syslog.syslog(syslog.LOG_ERR, "Failure: Shell Command")
        print(f"\n*********************\033[38;5;1m Shell Error \033[0;0m*********************")
        print(f"Command {e.args[1]} failed")
        print(f"Terminated by {command_result.returncode}")
        print(f"{command_result.stderr}")
        print("*******************************************************\n")
    except OSError as e:
        print(f"\n********************\033[38;5;1m Locked \033[0;0m**************************\n")
        print(f"Perhaps file is busy, locked, process blocked, or raced:\n") 
        print(f"{e}")
        print(f"*******************************************************\n")
    else:
        print(f"\n*********************\033[38;5;76m Success \033[0;0m*************************")
        print(f"\033[38;5;208mShell Errors: \033[0;0m {command_result.stdout}")
        print(f"*******************************************************\n")



# Set chmod for added convienence of the adminstrator with error handling and logging
@exception_handler
def setChmod(file, setting):
        os.fchmod(file, setting)
        writeLog("script", file + "was set to " + setting )




# Main
writeLog("script", date_time)
SetOpts("STARTUP")
SetOpts("SYSTEM")
SetOpts("KERNEL")
shellCommand("FILESEC")
shellCommand("USERSEC")


# Write succesfull completion to console and syslog
writeLog("script", "************ SUCCESS ************")
writeLog("script", "All files and directives validate")
writeLog("script", "*********************************")
writeLog("syslog", "SUCCESS: Hardening completed")

